### ISSUE TITLE
Slowdown of `tf.scan` operation in TF 2.5

### ISSUE THREAD

**Issue by dakshvar22** [tensorflow/tensorflow#51393]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS Mojave Version 10.14.5
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: -
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.5.0-rc3-213-ga4dfb8d1a71 2.5.0
- Python version: 3.7.5
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: -
- GPU model and memory: -

**Describe the current behavior**

There seems to be a considerable slowdown whenever a `tf.scan` operation is used inside the loss function of a model architecture and a learnable parameter (`self.transition_param` below) of the model architecture is part of the `tf.scan` op. Here's a code snippet to reproduce this:

```
import tensorflow as tf
import tensorflow_addons as tfa
from tqdm import tqdm
import numpy as np


def gen_batches(num_batches, batch_size, units):

    for _ in range(num_batches):
        x = np.random.random((batch_size, 20, units))
        y = np.random.randint(1, units, size=(batch_size, 20))
        yield x, y


class MyModel(tf.keras.models.Model):

    def __init__(self, units):
        super().__init__()

        self._tf_layers = {}

        self.units = units

        self.transition_param = self.add_weight(name="transition_param", shape=(units, units))

        self.optimizer = tf.keras.optimizers.Adam()
        self._training = False

    def _loss_fn_with_scan(self, inputs, transition_params):

        first_input = tf.slice(inputs, [0, 0, 0], [-1, 1, -1])
        first_input = tf.squeeze(first_input, [1])

        rest_of_input = tf.slice(inputs, [0, 1, 0], [-1, -1, -1])

        rest_of_input = tf.transpose(rest_of_input, [1, 0, 2])
        transition_params = tf.expand_dims(transition_params, 0)

        def _scan_fn(_state, _inputs):
            _state = tf.expand_dims(_state, 2)
            transition_scores = _state + transition_params
            new_alphas = _inputs + tf.reduce_logsumexp(transition_scores, [1])
            return new_alphas

        all_alphas = tf.transpose(tf.scan(_scan_fn, rest_of_input, first_input), [1, 0, 2])
        # add first state for sequences of length 1
        all_alphas = tf.concat([tf.expand_dims(first_input, 1), all_alphas], 1)

        return all_alphas

    def _loss(self, x, y):

        logits = tf.cast(x, dtype=tf.float32)

        loss = self._loss_fn_with_scan(logits, self.transition_param)

        return tf.reduce_mean(loss)

    @tf.function
    def train_on_batch(self, *args):
        with tf.GradientTape(persistent=True) as tape:
            loss = self._loss(*args)
        grads = tape.gradient(loss, self.trainable_weights)
        self.optimizer.apply_gradients(zip(grads, self.trainable_variables))
        return loss

    def train(self, epochs, batch_size, num_batches):

        data_generator_iter = gen_batches(num_batches, batch_size, self.units)

        sample_x, sample_y = next(data_generator_iter)

        self.train_on_batch(sample_x, sample_y)

        self._training = True

        progress_bar = tqdm(range(epochs), desc="Epochs")

        for epoch in progress_bar:
            for batch_x, batch_y in data_generator_iter:
                loss = self.train_on_batch(batch_x, batch_y)

            progress_bar.update(1)
            progress_bar.set_postfix({"loss": f"{loss.numpy():.3f}"})


num_batches = 5000
batch_size = 32
units = 64
epochs = 100

model = MyModel(units)
model.train(epochs, batch_size, num_batches)
```

The same code takes  ~ 3 mins, 16 seconds on TF 2.3 . However, it takes ~ 4 mins and 1 second on TF 2.5 .

The above code is a stripped down version of a model architecture which uses transformers + CRF layer. The `_loss_fn_with_scan` is taken from [`crf_log_norm` function](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/crf.py#L157) of tensorflow-addons package but as the snippet shows, the problem is already visible without the CRF bits of the code and just using `tf.scan` as part of the loss function (`tf.scan` is part of tensorflow repo, hence the issue seems appropriate here). On large datasets, the slowdown is considerably large. For example, with TF 2.3 it took 1 hour 20 mins to complete the training and with TF 2.5 it takes close to 3 hours to complete the training which is a considerable increase and a big blocker for us to upgrade to TF 2.5.

**Describe the expected behavior**

Training times should be comparable across TF 2.3 and TF 2.5 .

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): maybe
- Briefly describe your candidate solution(if contributing): I am not familiar with the internals of `tf.scan` op. If there's some help available, we could give it a try.

**Standalone code to reproduce the issue**
A snippet is available above.

**Comment by sanatmpa1** [comment]
I am able to reproduce the issue in colab with CPU runtime, it took 2 minutes 12 seconds with `TF 2.3` and 2 minutes 56 seconds for `TF 2.5`

Please find the [gist here](https://colab.research.google.com/gist/sanatmpa1/54bf06a243d227c67188c77f1f867ca0/51393-cpu.ipynb). Thanks!

**Comment by MarkDaoust** [comment]
Hi, 

There's a lot going on in that example code. 

Are you able to minimize the code and isolate the problem this a little more?

That will make it easier for anyone investigating to track down what's going on.

**Comment by dakshvar22** [comment]
Hi @MarkDaoust the code snippet has the minimal stuff needed to demonstrate the problem:

1. Dummy sample data
2. Custom model for a custom training loop
3. Cuatom loss function.
4. The loss function only has `tf.scan` op with a trainable param.

All of these are needed to demonstrate the problem.

Anything that you suggest to get rid of?

**Comment by zongweiz** [comment]
@sanatmpa1 @jvishnuvardhan Thanks for the repro. Could you please attach the full TensorFlow logs of the TF 2.3 run and TF 2.5 run, and also take the TensorFlow profile of a few training steps (https://www.tensorflow.org/guide/profiler#collect_performance_data)? This will help us isolate where the problem is (e.g., slow data processing, startup time, slow step time).

**Comment by koernerfelicia** [comment]
@sanatmpa1 @jvishnuvardhan is there any update on this? Anything we can do to help?

**Comment by bhack** [comment]
Can you reproduce this on TF 2.6 or TF nightly.
I've tried the Colab with TF 2.6 and It seems to freeze.

**Comment by sanatmpa1** [comment]
@bhack,

I've updated the [gist](https://colab.research.google.com/gist/sanatmpa1/54bf06a243d227c67188c77f1f867ca0/51393-cpu.ipynb#scrollTo=tgqM12gmpM5u) with results from TF 2.6 and TF nightly. Here's a comparison of results,

2.3      - `0:02:12.134876`
2.5      - `0:02:47.300976`
2.6      - `0:01:37.283745`
nightly -`0:02:46.074507`

**Comment by yarri-oss** [comment]
@sanatmpa1 are you able to respond to @zongweiz request for more profiling information in this gist? 

Here's an example of how you can add an [embedded TB profile run](https://colab.research.google.com/gist/yarri-oss/8d95ef510f6d1e88ef584d57617f57fb#scrollTo=TzD69-eA-K0R) to the 2.5 training. Thanks.

**Comment by bhack** [comment]
@sanatmpa1 I suppose that this numbers other then for 2.6 are wrong.
Can you add this in your Xolab to see if you have visible GPUs on all your tested versions?

```
physical_devices = tf.config.list_physical_devices('GPU')
print("Num GPUs:", len(physical_devices))
```

**Comment by dakshvar22** [comment]
@bhack @yarri-oss I can already see a small slowdown in a simple snippet with a single `tf.scan` operation.

Snippet:

```
import tensorflow as tf
from time import time
import numpy as np

start = time()

elems = np.random.random((20000))
sum = tf.scan(lambda a, x:
end = time()
print(end - start)
```

The snippet is taken from the [docs page](https://www.tensorflow.org/api_docs/python/tf/scan) of `tf.scan` op.

Runtime for TF 2.3: ~1.18s
Runtime for TF 2.5: ~1.3s

**System information**

Have I written custom code (as opposed to using a stock example script provided in TensorFlow): no
OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS Mojave Version 10.14.5
Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: -
TensorFlow installed from (source or binary): binary
TensorFlow version (use command below): v2.5.0-rc3-213-ga4dfb8d1a71 2.5.0
Python version: 3.7.5
Bazel version (if compiling from source):
GCC/Compiler version (if compiling from source):
CUDA/cuDNN version: -
GPU model and memory: -

This is a 10% slowdown which is almost in line with our larger experiments where we see an average slowdown of 18% in prediction time of our models using a `tf.scan` op internally. So, this is definitely one of the contributing factors to the overall slowdown.

Hope this is useful for further debugging and helps in making progress on it.

**Comment by bhack** [comment]
@dakshvar22 Can you test the same with TF nightly?

**Comment by dakshvar22** [comment]
@bhack `tf-nightly` is quite a lot slower! On the same system environment, the snippet runs in approximately 1.55s 🤯

**Comment by bhack** [comment]
Can you verify that you are placing the ops in the same device on all your tested version with this:

https://www.tensorflow.org/api_docs/python/tf/debugging/set_log_device_placement

**Comment by dakshvar22** [comment]
I don't have a GPU device on the system, so it should be using just the CPU device for all tested versions?

**Comment by dakshvar22** [comment]
@bhack I verified this by adding this line to the top of the snippet

```
tf.debugging.set_log_device_placement(True)
```

I see identical logs like this (truncated) on all versions:

```
....
2021-09-08 18:45:58.853542: I tensorflow/core/common_runtime/eager/execute.cc:733] Executing op LogicalAnd in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-08 18:45:58.853578: I tensorflow/core/common_runtime/eager/execute.cc:733] Executing op AddV2 in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-08 18:45:58.853653: I tensorflow/core/common_runtime/eager/execute.cc:733] Executing op LogicalAnd in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-08 18:45:58.853689: I tensorflow/core/common_runtime/eager/execute.cc:733] Executing op AddV2 in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-08 18:45:58.853751: I tensorflow/core/common_runtime/eager/execute.cc:733] Executing op LogicalAnd in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-08 18:45:58.853789: I tensorflow/core/common_runtime/eager/execute.cc:733] Executing op AddV2 in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-08 18:45:58.853851: I tensorflow/core/common_runtime/eager/execute.cc:733] Executing op LogicalAnd in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-08 18:45:58.853886: I tensorflow/core/common_runtime/eager/execute.cc:733] Executing op AddV2 in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-08 18:45:58.853983: I tensorflow/core/common_runtime/eager/execute.cc:733] Executing op LogicalAnd in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-08 18:45:58.870290: I tensorflow/core/common_runtime/eager/execute.cc:733] Executing op Pack in device /job:localhost/replica:0/task:0/device:CPU:0
```

**Comment by yarri-oss** [comment]
Hi - I am seeing improvements in TF 2.6 per this [gist](https://gist.github.com/yarri-oss/c3cd93f3c45606654bfa60241fd8e514), I agree it would be nice to add a `tf.scan` performance test to monitor any future regressions.

**Comment by koernerfelicia** [comment]
Hi @yarri-oss, I don't see this improvement, when I run all three versions 2.5 and 2.6 are slower than 2.3, see this [gist](https://gist.github.com/koernerfelicia/3c9a3b3ba6f23914afe9cec309a6d41b#file-notebook-ipynb)

**Comment by bhack** [comment]
I've tested this on my local system with  TF-nighlty TF 2.6.0 TF 2.5.0:

```python
import tensorflow as tf
from time import time
import numpy as np
from pyinstrument import Profiler

print("version {}".format(tf.__version__))
np.random.seed(0) 
input_test = np.random.random((20000))
for i in range(30):
    tf.scan(lambda a, x: a+x, input_test)
```
You can see here some of the core operations:

![immagine](https://user-images.githubusercontent.com/1710528/133339191-c35d5910-6f18-4ec4-a3fe-c0c42d00a739.png)

I don't see too much difference between 2.5 and 2.6. I see a slower `add_v2`  in  `nightly` and probably a little bit of overhead with the new `error_handler` filtering machinery that we have introduced in master.

**Comment by koernerfelicia** [comment]
@bhack, have you compared this to 2.3? I also don't see much difference between 2.5 and 2.6, but a significant difference from 2.3 to 2.5

**Comment by bhack** [comment]
I don't have the env ready for 2.3 now but you can post your graph with https://pyinstrument.readthedocs.io/en/latest/.
You need to use `-r html` for the interactive output so the you can take a screenshot of the critical section as I've done in the previous post.

**Comment by koernerfelicia** [comment]
2.3:
<img width="932" alt="Screenshot 2021-09-29 at 09 02 49" src="https://user-images.githubusercontent.com/45405119/135219257-67fdcd1d-f404-4133-bb25-3f277b2033e0.png">

html [here](https://gistpreview.github.io/?31109fd77a36bd4722e52e271d2998c5)

2.5:
<img width="938" alt="2_5" src="https://user-images.githubusercontent.com/45405119/135220112-98eacf4b-0be8-49ec-9b37-983ad7757921.png">

html [here](https://gistpreview.github.io/?6c8af2a23a0a6d56bad2e60e16d5b924)

2.6:
<img width="934" alt="2_6" src="https://user-images.githubusercontent.com/45405119/135220119-e6ff902d-242f-427f-be90-d112eb62f6a9.png">

html [here](https://gistpreview.github.io/?fb86e9e63ba8f55ed04b9b12e88b9a0c)

**Comment by koernerfelicia** [comment]
>Hi @yarri-oss, I don't see this improvement, when I run all three versions 2.5 and 2.6 are slower than 2.3, see this gist

@yarri-oss did you get a chance to look at the gist I shared above?

**Comment by bhack** [comment]
Thanks, yes in eager mode it seems that `add_v2` is slower than `2.3`. Is this kernel using MLIR now?

**Comment by jsmeredith** [comment]
Hi, all!  Just wanted to let you know we're investigating -- no conclusions yet, but I have a couple leads I'm following.  I'll post back when I have more.  Thanks for the info so far!

**Comment by saxenasaurabh** [comment]
Do you still see a regression if you don't use `persistent=True` (doesn't seem like you are computing any higher order gradients so functionally your model should be equivalent)?

**Comment by markus-hinsche** [comment]
Hi everybody, I work with @dakshvar22 and @koernerfelicia on the problem and have done some measurements on CPU:

![image](https://user-images.githubusercontent.com/1809832/142467192-87ad7618-9441-4891-99f1-5b9f21140a92.png)

Here are some explanations about this image:
* The meaning of the **env**  column is mostly the tensorflow version:
    * 2.3-env: TF 2.3
    * 2.5-env: TF 2.5
    * 2.6-env: TF 2.6
    * 2.6-env-new: TF. 2.6 with the suggested gradient tape `persistent=False` change (https://github.com/RasaHQ/rasa/pull/9867)
* `carbon-bot` and `financial-demo` are just 2 different internal dataset I measured on.
* We measured on the different configs that might be affected by the slowdown
* You can see the measured training time (columns: `total train`, `diet train`) in seconds
* The relative speedups are just inferred values, they correspond to the respective row for `2.3-env`

What we learn from this:
1. The change brings training times back to what we had for TF 2.3 version
2. However, this should not be considered a fix of the original root cause of the problem.
3. Inference time is still taking a hit. This is most likely because of the `tf.scan` operation being slow

Hope this helps with your investigation @jvishnuvardhan @jsmeredith . We are also planning to do some measurements on GPU

**Comment by dakshvar22** [comment]
Hi @jsmeredith .. Is this still something you are looking into? Did you get a chance to look at the [above observations](https://github.com/tensorflow/tensorflow/issues/51393#issuecomment-973114987)?

**Comment by jsmeredith** [comment]
Greetings!  Sure, let me take another look -- I wasn't entirely sure if I was awaiting GPU data .  (I also admit I'm making heavy use of other folks' expertise on this one, and alas some of those people are now on different teams over the holidays.)  The timing is good though, I'm just wrapping up one project so I have a little time to see what we can do here.

**Comment by koernerfelicia** [comment]
Hi @jsmeredith, do you have any updates on this issue? Would it be helpful to get measurements on GPU?

**Comment by sachinprasadhs** [comment]
I tested with Tensorflow 2.12 version and I'm able to notice improvement in performance, the time taken is 1 minute and 15 seconds. 
Please find the attached Gist [here](https://gist.github.com/sachinprasadhs/9ade4cbb0e003d06f13ab80ce1004004) for reference. Thanks!
